lookout/style/format/research/robustness.py

```diff
@@ -0,0 +1,146 @@
+from argparse import ArgumentParser
+from collections import defaultdict
+from difflib import SequenceMatcher
+import glob
+import os
+from typing import Callable, Dict, List, NamedTuple, Set, Tuple
+
+from bblfsh import BblfshClient
+import numpy
+
+from lookout.core.slogging import setup
+from lookout.style.format.feature_extractor import FeatureExtractor
+from lookout.style.format.feature_utils import CLASS_INDEX
+from lookout.style.format.files_filtering import filter_filepaths
+from lookout.style.format.model import FormatModel
+from lookout.style.format.quality_report import prepare_files
+
+
+Misprediction = NamedTuple("Misprediction", [("y", numpy.ndarray), ("pred", numpy.ndarray),
+                                             ("node", List[Callable]), ("rule", numpy.ndarray)])
+
+
+def get_content_from_repo(folder: str) -> Dict[str, str]:
+    content = {}
+    filenames = glob.glob(folder, recursive=True)
+    for file in filter_filepaths(filenames):
+        with open(file) as g:
+            content[file] = g.read()
+    return content
+
+    
+def get_difflib_changes(true_content: Dict[str, str], noisy_content: Dict[str, str]
+                        ) -> Tuple[List[str], List[str], Dict[str, Set[int]], int]:
+    true_files, noisy_files = set(), set()
+    nb_changes = 0
+    lines_changed = defaultdict(set)
+    for (tf, tc), (nf, nc) in zip(true_content.items(), noisy_content.items()):
+        matcher = SequenceMatcher(a=tc.splitlines(), b=nc.splitlines())
+        for action, _, _, j1, j2 in matcher.get_opcodes():
+            if action is not "equal":
+                lines_changed[nf].update(range(j1, j2+1))
+                true_files.add(tf)
+                noisy_files.add(nf)
+                nb_changes += 1
+    return sorted(list(true_files)), sorted(list(noisy_files)), lines_changed, nb_changes
+
+
+def get_mispreds(y: numpy.ndarray, y_pred: numpy.ndarray, nodes: List[Callable],
+                 winner: numpy.ndarray) -> List[NamedTuple]:
+    mispreds = []
+    for gt, pred, vn, rule in zip(y, y_pred, nodes, winner):
+        if gt != pred:
+            mispreds.append(Misprediction(gt, pred, vn, rule))
+    return mispreds
+
+
+def get_diff_mispreds(mispreds_noise: List[NamedTuple], lines_changed: Dict[str, Set[int]]
```

Comment:
It's better to use the broadest possible type in function argument and the narrowest possible type in return types, so here it would be Iterable[NamedTuple] and Mapping[str, Set[int]] for the argument types